common/file_watcher_hub.cpp

```diff
@@ -0,0 +1,91 @@
+/// Copyright 2019 Pinterest Inc.
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+/// http://www.apache.org/licenses/LICENSE-2.0
+
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+
+#include "file_watcher.h"
+#include "file_watcher_hub.h"
+#include "glog/logging.h"
+
+namespace common {
+
+FileWatcherHub * FileWatcherHub::Instance() {
+  static FileWatcherHub instance;
+  return &instance;
+}
+
+typedef std::pair<std::string, EventObserverPtr> ObserverEventCallbackPair;
+typedef std::pair<std::string, IdToObserverMapPtr> FileNameObserversMapPair;
+
+bool FileWatcherHub::subscribe(
+  std::string fileName,
+  std::string observerId,
+  EventObserverPtr observerPtr) {
+  folly::RWSpinLock::WriteHolder write_gaurd(subscription_lock_);
+  auto watcherIter = subscriptions_.find(fileName);
+  if (watcherIter == subscriptions_.end()) {
+    // Add to the subscription
+    IdToObserverMapPtr idToObserverMapPtr = std::make_shared<IdToObserverMap>();
+    subscriptions_.insert(FileNameObserversMapPair(fileName, idToObserverMapPtr));
+    watcherIter = subscriptions_.find(fileName);
+
+    auto obsPtr = watcherIter->second->find(observerId);
+    if (obsPtr == watcherIter->second->end()) {
+      watcherIter->second->insert(ObserverEventCallbackPair(observerId, observerPtr));
+    }
+
+    CHECK(common::FileWatcher::Instance()->AddFile(
+      fileName,
+      [fileName = fileName, this] (std::string content) {
+        folly::RWSpinLock::ReadHolder read_gaurd(subscription_lock_);
```

Comment:
This is a deadlock here...

FileWatcher AddFile will immediately callback ones a successful watch is established... which means... here we will have FileWatcher thread calling this callback trying to obtain subscription_lock_ ... which is already held by thread calling subscribe ...